File: kartwheel/app/src/main/java/us/handstand/kartwheel/model/Event.java
package us.handstand.kartwheel.model;


import android.content.ContentValues;
import android.database.sqlite.SQLiteDatabase;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.google.auto.value.AutoValue;
import com.google.gson.Gson;
import com.google.gson.TypeAdapter;
import com.squareup.sqlbrite.BriteDatabase;

import java.util.Date;

import us.handstand.kartwheel.model.columnadapter.ColumnAdapters;

import static us.handstand.kartwheel.model.Util.putIfNotAbsent;

@AutoValue
public abstract class Event implements EventModel {

    public static final Factory<Event> FACTORY = new Factory<>(new Creator<Event>() {
        @Override
        public Event create(@NonNull String id, @NonNull Date endTime, @NonNull String name, @NonNull Date startTime, @NonNull String updatedAt, @Nullable Boolean usersCanSeeRaces) {
            return new AutoValue_Event(id, endTime, name, startTime, updatedAt, usersCanSeeRaces);
        }
    }, ColumnAdapters.DATE_LONG, ColumnAdapters.DATE_LONG);

    public void insert(@Nullable BriteDatabase db) {
        if (db != null) {
            ContentValues cv = new ContentValues();
            putIfNotAbsent(cv, ID, id());
            putIfNotAbsent(cv, ENDTIME, endTime().getTime());
            putIfNotAbsent(cv, NAME, name());
            putIfNotAbsent(cv, STARTTIME, startTime().getTime());
            putIfNotAbsent(cv, UPDATEDAT, updatedAt());
            putIfNotAbsent(cv, USERSCANSEERACES, usersCanSeeRaces());
            db.insert(TABLE_NAME, cv, SQLiteDatabase.CONFLICT_REPLACE);
        }
    }

    // Required by Gson
    @SuppressWarnings("unused")
    public static TypeAdapter<Event> typeAdapter(Gson gson) {
        return new AutoValue_Event.GsonTypeAdapter(gson);
    }
}

File: kartwheel/app/src/main/kotlin/us/handstand/kartwheel/model/Event.kt
package us.handstand.kartwheel.model


import android.content.ContentValues
import com.google.auto.value.AutoValue
import com.google.gson.Gson
import com.google.gson.TypeAdapter
import us.handstand.kartwheel.model.EventModel.Creator
import us.handstand.kartwheel.model.Util.putIfNotAbsent

@AutoValue
abstract class Event : EventModel, Insertable {

    override fun tableName(): String {
        return EventModel.TABLE_NAME
    }

    override val contentValues: ContentValues
        get() {
            val cv = ContentValues()
            putIfNotAbsent(cv, EventModel.ID, id())
            putIfNotAbsent(cv, EventModel.ENDTIME, endTime().time)
            putIfNotAbsent(cv, EventModel.NAME, name())
            putIfNotAbsent(cv, EventModel.STARTTIME, startTime().time)
            putIfNotAbsent(cv, EventModel.UPDATEDAT, updatedAt()?.time)
            putIfNotAbsent(cv, EventModel.USERSCANSEERACES, usersCanSeeRaces())
            return cv
        }

    companion object : Creator<Event> by Creator(::AutoValue_Event) {
        val FACTORY = EventModel.Factory<Event>(Creator<Event> { id, endTime, name, startTime, updatedAt, usersCanSeeRaces -> create(id, endTime, name, startTime, updatedAt, usersCanSeeRaces) }, ColumnAdapters.DATE_LONG, ColumnAdapters.DATE_LONG, ColumnAdapters.DATE_LONG)
        // Required by Gson
        @JvmStatic
        fun typeAdapter(gson: Gson): TypeAdapter<Event> {
            return AutoValue_Event.GsonTypeAdapter(gson)
        }
    }
}
